android/app/src/main/kotlin/coop/polypoly/polypod/updatenotification/UpdateNotification.kt

```diff
@@ -5,23 +5,65 @@ import coop.polypoly.polypod.Preferences
 import coop.polypoly.polypod.R
 
 class UpdateNotification(private val context: Context) {
-    val id = context.resources.getInteger(R.integer.update_notification_id)
+    companion object {
+        data class MockData(var id: Int? = null)
+        val mockData = MockData()
+    }
+
+    private enum class State {
+        NOT_SEEN,
+        PUSH_SEEN,
+        ALL_SEEN;
+
+        companion object {
+            fun parse(s: String?) =
+                s.let { name -> values().firstOrNull { it.name == name } }
+        }
+    }
+
+    val id = mockData.id
+        ?: context.resources.getInteger(R.integer.update_notification_id)
+
     val title = context.getString(R.string.update_notification_title)
+
     val text = context.getString(R.string.update_notification_text)
+
     val pushDelay =
         context.resources.getInteger(R.integer.update_notification_push_delay)
 
-    val inAppNotificationSeen: Boolean
-        get() = id == 0 || Preferences.getSeenInAppNotificationId(context) == id
+    val showPush: Boolean
+        get() = state == State.NOT_SEEN
+
+    val showInApp: Boolean
+        get() = state != State.ALL_SEEN
 
-    val pushNotificationSeen: Boolean
-        get() = id == 0 || Preferences.getSeenPushNotificationId(context) == id
+    private var state: State = loadLastState()
+        set(value) {
+            field = value
+            Preferences.setLastNotification(context, id, state.name)
+        }
+
+    private fun loadLastState(): State {
+        val (lastId, lastState) = Preferences.getLastNotification(context)
+        return when (id) {
+            0 -> State.ALL_SEEN
+            lastId -> State.parse(lastState) ?: State.ALL_SEEN
+            else -> State.NOT_SEEN
+        }
+    }
 
-    fun markInAppNotificationSeen() {
-        Preferences.setSeenInAppNotificationId(context, id)
+    fun onFirstRun() {
```

Comment:
Wasn't really sure about the naming of these event handlers starting with `on`, ideas welcome :D

Reply:
Why not the initial `mark`? Or more directly `setStateToAllSeen`?

Reply:
Well, I didn't just want to expose the state machine, I wanted to abstract the logic away - only `UpdateNotification` should worry about what happens on startup, on first run, etc.

Reply:
The problem is that one expects `on` stuff to be fired kinda automatically, when `xxx` happens. So maybe use `at` or `when`? Major naming nit, I know.

Reply:
Well, I was asking for better naming here, so... I suppose `on` is a rather common naming scheme for event handlers (in the JS world at least), but on I could also go for `handle`, or even `trigger`.

I kinda like trigger, but it doesn't actually _trigger a startup_ for example, so maybe we'll just leave it as is for now and think of better names next time we touch this code :D

Reply:
Ah screw it, I changed it to `handle`, it's a verb which is always nice, and I figure the code is a little bit more clear this way.